scripts/ensure_exports_are_documented.py

```diff
@@ -41,19 +42,22 @@ def iter_all_documented_names() -> t.Iterable[str]:
     # names under these directives
     #
     #   .. autoclass:: <name>
+    #   .. autoclass:: <name>(<args>)
     #   .. autofunction:: <name>
     #   .. autoexception:: <name>
     #   .. autodata:: <name>
+    auto_directive = r"auto(?:class|function|exception|data)"
+    name_capturer = r"(?:\w+\.)*(\w+)(?:\(.*\))?"
     autodoc_pattern = re.compile(
-        r"^\.\.\s+auto(?:class|function|exception|data)\:\:\s+(?:\w+\.)*(\w+)$",
+        rf"^\.\.\s+{auto_directive}::\s+{name_capturer}$",
```

Comment:
May I suggest instead of creating a `rf` string, that we switch to a verbose regex and embed any necessary comments?

I think this is harder to read for breaking the expression up into multiple assembled strings, but I understand the impulse to break it into multiple physical lines.

Reply:
As a separate matter, I don't understand why it's desirable to allow for the empty `()` after a classname.
I would rather see
```rst
.. autoclass:: Noodle
```
than
```rst
.. autoclass:: Noodle()
```

simply on the grounds that the bare name is the norm in sphinx documentation. I didn't even know until a few weeks ago that the empty args were _valid_, that's how unusual it is in my experience!

Could you share why you are including the empty args to classnames in autoclass directives?

Reply:
> and embed any necessary comments

Not sure what you mean by this. Like embed the comments within the regex?

Reply:
> As a separate matter, I don't understand why it's desirable to allow for the empty () after a classname.

Omitting the `()` instructs sphinx to auto-insert init args into the generated docs page.  Adding a `()` tells spinx that "I will explicitly list out the parameters to this class's init".

In these cases I used the empty `()` to tell sphinx not to list out any init-args. That shouldn't be the default there are cases where it's valuable to have that ability; specifically:
 * Abstract base classes that a user shouldn't be instantiating (`GlobusApp`).
 * Data classes where the parameters are listed more visibly as ivars (`GlobusAppConfig`).
 * Protocols. (Providers and handlers in this case)

Here's some side-by-sides of what gets generated in those different cases that hopefully demonstrate why I opted to pull redundant or unnecessary info that in some cases was taking up a good amount of screen space.

### GlobusApp

In this case there's 5 added lines of terse arg listing that will be skipped over basically every time someone reads this page; so better to omit it.

**With `()`**

<img width="815" alt="image" src="https://github.com/user-attachments/assets/36f75bec-8099-49c2-abc7-89198f2ca872">

**Without `()`**

<img width="827" alt="image" src="https://github.com/user-attachments/assets/7b6e7bda-06f6-4b35-b185-f0b8c4e2777c">


### GlobusAppConfig

**With `()`**

<img width="806" alt="image" src="https://github.com/user-attachments/assets/06422097-ced2-4fef-b8e9-1a85ea70d1bc">


** Without `()`**

<img width="796" alt="image" src="https://github.com/user-attachments/assets/1504466e-a1d8-4a82-84bb-770a8ba9db74">


### Protocols

**With `()`**

<img width="839" alt="image" src="https://github.com/user-attachments/assets/5ad97601-8e05-4498-ab71-c5ee0933095b">


** Without `()`**

<img width="780" alt="image" src="https://github.com/user-attachments/assets/3ff8cc28-0685-40e4-b6e9-4138ba16a23d">


---

Long message, but it's important to look at what documentation we're actually generating and choose what'll be most helpful there not just what looks cleanest in the rst format; and I hold that each of these instances have a good justification for omitting those init params.

Reply:
Ooh, okay, I didn't know that `()` actually had an impact on Sphinx behavior.

I agree with the protocol case, where the `*args, **kwargs` inclusion is basically meaningless.

I disagree with the others, but I'm not sure I want to block this PR on resolving our difference of opinion on this topic.
The initializer for a class -- including for an abstract class -- is one of its interfaces.

That said, I hate the generated content for `GlobusApp`'s signature, since it expands the config. I would like best
```
class globus_sdk.GlobusApp(
    app_name='Unnamed Globus App',
    *,
    login_client=None,
    client_id=None,
    client_secret=None,
    scope_requirements=None,
    config=<DEFAULT_CONFIG>
)
```

(Aside: I looked into getting `furo` to generate this kind of spacing, and learned that it's basically infeasible unless something changes in the generated sphinx html.)

---

I'm happy to accept this for now, and if I feel that we need to document `__init__` for a class separately, I can submit my follow-up for us to discuss.

Reply:
My strength of opinion is:
 * Strongest: protocols
 * Stronger: GlobusApp
 * Medium Strength: Data Class

But we can resolve this discussion for now if you're happy enough as-is & schedule some time in the arena to duke it out;

Reply:
Updated regex back to inline w/ `re.X`